File: ribo_switch/energy.py
import math
from ribo_switch.types import Base, Energy, Sequence, Structure, HairpinLoop, StackLoop, InteriorLoop, BulgeLoop, MultiLoop, ExternalLoop, LoopType, CANONICAL_PAIRS
from ribo_switch.structure import decompose_loops
from ribo_switch.turner import TurnerParams, INF

def eval_energy(seq: Sequence, structure: Structure, params: TurnerParams) -> Energy:
    if len(seq) != structure.length:
        raise ValueError(f'Sequence length ({len(seq)}) != structure length ({structure.length})')
    loops = decompose_loops(structure)
    total = 0
    for loop in loops:
        total += _loop_energy(seq, loop, params)
    return total

def _loop_energy(seq: Sequence, loop: LoopType, params: TurnerParams) -> Energy:
    if isinstance(loop, HairpinLoop):
        return hairpin_energy(seq, loop.closing_pair, params)
    elif isinstance(loop, StackLoop):
        return stack_energy(seq, loop.outer_pair, loop.inner_pair, params)
    elif isinstance(loop, InteriorLoop):
        return interior_energy(seq, loop.outer_pair, loop.inner_pair, len(loop.left_unpaired), len(loop.right_unpaired), params)
    elif isinstance(loop, BulgeLoop):
        return bulge_energy(seq, loop.outer_pair, loop.inner_pair, len(loop.unpaired), params)
    elif isinstance(loop, MultiLoop):
        return multiloop_energy(seq, loop.closing_pair, loop.branches, len(loop.unpaired), params)
    elif isinstance(loop, ExternalLoop):
        return external_energy(seq, loop, params)
    else:
        raise TypeError(f'Unknown loop type: {type(loop)}')

def _pair_index(seq: Sequence, i: int, j: int) -> int:
    bi, bj = (seq.bases[i], seq.bases[j])
    _PAIR_MAP = {(Base.A, Base.U): 0, (Base.U, Base.A): 1, (Base.C, Base.G): 2, (Base.G, Base.C): 3, (Base.G, Base.U): 4, (Base.U, Base.G): 5}
    idx = _PAIR_MAP.get((bi, bj))
    if idx is None:
        raise ValueError(f'Non-canonical pair ({bi.name}, {bj.name}) at ({i}, {j})')
    return idx

def hairpin_energy(seq: Sequence, closing: tuple[int, int], params: TurnerParams) -> Energy:
    i, j = closing
    size = j - i - 1
    if size < 3:
        return INF
    if size <= 30:
        energy = int(params.hairpin_init[size])
    else:
        energy = int(params.hairpin_init[30]) + int(round(params.loop_extrapolation_coeff * 100 * math.log(size / 30.0)))
    pair_idx = _pair_index(seq, i, j)
    if size == 3:
        loop_seq = ''.join((seq.bases[k].name for k in range(i, j + 1)))
        bonus = params.hairpin_triloop.get(loop_seq, 0)
        if bonus:
            energy = bonus
        if pair_idx in (0, 1, 4, 5):
            energy += params.terminal_au_penalty
        if all((seq.bases[k] == Base.C for k in range(i + 1, j))):
            energy += params.hairpin_c3
        return energy
    if size == 4:
        loop_seq = ''.join((seq.bases[k].name for k in range(i, j + 1)))
        bonus = params.hairpin_tetraloop.get(loop_seq, 0)
        if bonus:
            return bonus
    b5 = seq.bases[i + 1].value
    b3 = seq.bases[j - 1].value
    energy += int(params.hairpin_mismatch[pair_idx][b5][b3])
    first_mm = (seq.bases[i + 1], seq.bases[j - 1])
    if first_mm in ((Base.U, Base.U), (Base.G, Base.A)):
        energy += params.hairpin_uu_ga_bonus
    if first_mm == (Base.G, Base.G):
        energy += params.hairpin_gg_bonus
    if pair_idx == 4:
        if i >= 2 and seq.bases[i - 1] == Base.G and (seq.bases[i - 2] == Base.G):
            energy += params.hairpin_special_gu
    if all((seq.bases[k] == Base.C for k in range(i + 1, j))):
        energy += params.hairpin_c_slope * size + params.hairpin_c_intercept
    return energy

def stack_energy(seq: Sequence, outer: tuple[int, int], inner: tuple[int, int], params: TurnerParams) -> Energy:
    outer_idx = _pair_index(seq, outer[0], outer[1])
    inner_idx = _pair_index(seq, inner[0], inner[1])
    return int(params.stack[outer_idx][inner_idx])

def interior_energy(seq: Sequence, outer: tuple[int, int], inner: tuple[int, int], n_left: int, n_right: int, params: TurnerParams) -> Energy:
    i, j = outer
    p, q = inner
    outer_idx = _pair_index(seq, i, j)
    inner_idx = _pair_index(seq, p, q)
    total_unpaired = n_left + n_right
    if n_left == 1 and n_right == 1:
        mm5 = seq.bases[i + 1].value
        mm3 = seq.bases[j - 1].value
        val = int(params.interior_1x1[outer_idx][inner_idx][mm5][mm3])
        if val < INF:
            return val
    if total_unpaired > 30:
        energy = int(params.interior_init[30]) + int(round(params.loop_extrapolation_coeff * 100 * math.log(total_unpaired / 30.0)))
    else:
        energy = int(params.interior_init[total_unpaired])
    asymmetry = abs(n_left - n_right)
    ninio = min(params.ninio_max, params.ninio_m * asymmetry)
    energy += ninio
    if not (n_left == 1 and n_right == 1):
        b5_outer = seq.bases[i + 1].value
        b3_outer = seq.bases[j - 1].value
        energy += int(params.interior_mismatch[outer_idx][b5_outer][b3_outer])
        b5_inner = seq.bases[q + 1].value if q + 1 < len(seq) else 0
        b3_inner = seq.bases[p - 1].value if p - 1 >= 0 else 0
        energy += int(params.interior_mismatch[inner_idx][b3_inner][b5_inner])
    if outer_idx in (0, 1, 4, 5):
        energy += params.terminal_au_penalty
    if inner_idx in (0, 1, 4, 5):
        energy += params.terminal_au_penalty
    return energy

def bulge_energy(seq: Sequence, outer: tuple[int, int], inner: tuple[int, int], n_unpaired: int, params: TurnerParams) -> Energy:
    if n_unpaired > 30:
        energy = int(params.bulge_init[30]) + int(round(params.loop_extrapolation_coeff * 100 * math.log(n_unpaired / 30.0)))
    else:
        energy = int(params.bulge_init[n_unpaired])
    outer_idx = _pair_index(seq, outer[0], outer[1])
    inner_idx = _pair_index(seq, inner[0], inner[1])
    if n_unpaired == 1:
        energy += int(params.stack[outer_idx][inner_idx])
    if outer_idx in (0, 1, 4, 5):
        energy += params.terminal_au_penalty
    if inner_idx in (0, 1, 4, 5):
        energy += params.terminal_au_penalty
    return energy

def multiloop_energy(seq: Sequence, closing: tuple[int, int], branches: list[tuple[int, int]], n_unpaired: int, params: TurnerParams) -> Energy:
    n_branches = len(branches)
    energy = params.ml_offset + params.ml_per_branch * (n_branches + 1) + params.ml_per_unpaired * n_unpaired
    closing_idx = _pair_index(seq, closing[0], closing[1])
    if closing_idx in (0, 1, 4, 5):
        energy += params.terminal_au_penalty
    for p, q in branches:
        branch_idx = _pair_index(seq, p, q)
        if branch_idx in (0, 1, 4, 5):
            energy += params.terminal_au_penalty
    return energy

def external_energy(seq: Sequence, loop: ExternalLoop, params: TurnerParams) -> Energy:
    energy = 0
    n = len(seq)
    for i, j in loop.closing_pairs:
        pair_idx = _pair_index(seq, i, j)
        if pair_idx in (0, 1, 4, 5):
            energy += params.terminal_au_penalty
        if i > 0:
            d5_base = seq.bases[i - 1].value
            energy += int(params.dangle5[pair_idx][d5_base])
        if j + 1 < n:
            d3_base = seq.bases[j + 1].value
            energy += int(params.dangle3[pair_idx][d3_base])
    return energy

File: ribo_rs/src/energy.rs
use crate::params::*;
pub fn eval_energy(seq: &[u8], pair_table: &[i32]) -> i32 {
    let n = seq.len();
    assert_eq!(n, pair_table.len());
    let mut total = 0i32;
    let mut top_pairs: Vec<(usize, usize)> = Vec::new();
    let mut pos = 0usize;
    while pos < n {
        let j = pair_table[pos];
        if j > pos as i32 {
            top_pairs.push((pos, j as usize));
            pos = j as usize + 1;
        } else {
            pos += 1;
        }
    }
    total += external_energy(seq, pair_table, &top_pairs, n);
    for &(i, j) in &top_pairs {
        total += process_pair(seq, pair_table, i, j);
    }
    total
}
fn process_pair(seq: &[u8], pair_table: &[i32], i: usize, j: usize) -> i32 {
    let children = find_children(pair_table, i, j);
    let mut total = match children.len() {
        0 => hairpin_energy(seq, i, j),
        1 => {
            let (p, q) = children[0];
            let nl = p - i - 1;
            let nr = j - q - 1;
            if nl == 0 && nr == 0 {
                stack_energy(seq, i, j, p, q)
            } else if nl == 0 || nr == 0 {
                bulge_energy(seq, i, j, p, q, nl + nr)
            } else {
                interior_energy(seq, i, j, p, q, nl, nr)
            }
        }
        _ => multiloop_energy(seq, pair_table, i, j, &children),
    };
    for &(p, q) in &children {
        total += process_pair(seq, pair_table, p, q);
    }
    total
}
fn find_children(pair_table: &[i32], i: usize, j: usize) -> Vec<(usize, usize)> {
    let mut children = Vec::new();
    let mut pos = i + 1;
    while pos < j {
        let q = pair_table[pos];
        if q > pos as i32 && (q as usize) < j {
            children.push((pos, q as usize));
            pos = q as usize + 1;
        } else {
            pos += 1;
        }
    }
    children
}
fn hairpin_energy(seq: &[u8], i: usize, j: usize) -> i32 {
    let size = j - i - 1;
    if size < MIN_HAIRPIN {
        return INF;
    }
    let pi = match pair_index(seq[i], seq[j]) {
        Some(p) => p,
        None => return INF,
    };
    let mut energy = if size <= 30 {
        HAIRPIN_INIT[size]
    } else {
        HAIRPIN_INIT[30]
            + (LOOP_EXTRAPOLATION_COEFF * (size as f64 / 30.0).ln() * 100.0).round() as i32
    };
    if size == 3 {
        let w = &seq[i..=j];
        let bonus = triloop_bonus(w);
        if bonus != 0 {
            energy = bonus;
        }
        if is_au_gu(pi) {
            energy += TERMINAL_AU_PENALTY;
        }
        if (i + 1..j).all(|k| seq[k] == 1) {
            energy += HAIRPIN_C3;
        }
        return energy;
    }
    if size == 4 {
        let w = &seq[i..=j];
        let bonus = tetraloop_bonus(w);
        if bonus != 0 {
            return bonus;
        }
    }
    let b5 = seq[i + 1] as usize;
    let b3 = seq[j - 1] as usize;
    energy += HAIRPIN_MM[pi][b5][b3];
    let mm = (seq[i + 1], seq[j - 1]);
    if mm == (3, 3) || mm == (2, 0) {
        energy += HAIRPIN_UU_GA_BONUS;
    }
    if mm == (2, 2) {
        energy += HAIRPIN_GG_BONUS;
    }
    if pi == 4 && i >= 2 && seq[i - 1] == 2 && seq[i - 2] == 2 {
        energy += HAIRPIN_SPECIAL_GU;
    }
    if (i + 1..j).all(|k| seq[k] == 1) {
        energy += HAIRPIN_C_SLOPE * size as i32 + HAIRPIN_C_INTERCEPT;
    }
    energy
}
fn stack_energy(seq: &[u8], i: usize, j: usize, p: usize, q: usize) -> i32 {
    let oi = match pair_index(seq[i], seq[j]) { Some(x) => x, None => return INF };
    let ii = match pair_index(seq[p], seq[q]) { Some(x) => x, None => return INF };
    STACK[oi][ii]
}
fn bulge_energy(seq: &[u8], i: usize, j: usize, p: usize, q: usize, size: usize) -> i32 {
    let mut energy = if size <= 30 {
        BULGE_INIT[size]
    } else {
        BULGE_INIT[30]
            + (LOOP_EXTRAPOLATION_COEFF * (size as f64 / 30.0).ln() * 100.0).round() as i32
    };
    let oi = match pair_index(seq[i], seq[j]) { Some(x) => x, None => return INF };
    let ii = match pair_index(seq[p], seq[q]) { Some(x) => x, None => return INF };
    if size == 1 {
        energy += STACK[oi][ii];
    }
    if is_au_gu(oi) { energy += TERMINAL_AU_PENALTY; }
    if is_au_gu(ii) { energy += TERMINAL_AU_PENALTY; }
    energy
}
fn interior_energy(
    seq: &[u8],
    i: usize, j: usize,
    p: usize, q: usize,
    nl: usize, nr: usize,
) -> i32 {
    let oi = match pair_index(seq[i], seq[j]) { Some(x) => x, None => return INF };
    let ii = match pair_index(seq[p], seq[q]) { Some(x) => x, None => return INF };
    let total = nl + nr;
    if nl == 1 && nr == 1 {
        let mm5 = seq[i + 1] as usize;
        let mm3 = seq[j - 1] as usize;
        let val = INT11[oi][ii][mm5][mm3];
        if val < INF {
            return val;
        }
    }
    let mut energy = if total <= 30 {
        INTERIOR_INIT[total]
    } else {
        INTERIOR_INIT[30]
            + (LOOP_EXTRAPOLATION_COEFF * (total as f64 / 30.0).ln() * 100.0).round() as i32
    };
    let asym = (nl as i32 - nr as i32).unsigned_abs() as i32;
    energy += (NINIO_M * asym).min(NINIO_MAX);
    if !(nl == 1 && nr == 1) {
        let b5o = seq[i + 1] as usize;
        let b3o = seq[j - 1] as usize;
        energy += INTERIOR_MM[oi][b5o][b3o];
        let b3i = if p > 0 { seq[p - 1] as usize } else { 0 };
        let b5i = if q + 1 < seq.len() { seq[q + 1] as usize } else { 0 };
        energy += INTERIOR_MM[ii][b3i][b5i];
    }
    if is_au_gu(oi) { energy += TERMINAL_AU_PENALTY; }
    if is_au_gu(ii) { energy += TERMINAL_AU_PENALTY; }
    energy
}
fn multiloop_energy(
    seq: &[u8],
    _pair_table: &[i32],
    i: usize, j: usize,
    children: &[(usize, usize)],
) -> i32 {
    let nb = (children.len() as i32) + 1;
    let mut n_unpaired = 0i32;
    let mut prev = i;
    for &(p, q) in children {
        n_unpaired += (p - prev - 1) as i32;
        prev = q;
    }
    n_unpaired += (j - prev - 1) as i32;
    let mut energy = ML_OFFSET + ML_PER_BRANCH * nb + ML_PER_UNPAIRED * n_unpaired;
    let ci = match pair_index(seq[i], seq[j]) { Some(x) => x, None => return INF };
    if is_au_gu(ci) { energy += TERMINAL_AU_PENALTY; }
    for &(p, q) in children {
        if let Some(bi) = pair_index(seq[p], seq[q]) {
            if is_au_gu(bi) { energy += TERMINAL_AU_PENALTY; }
        }
    }
    energy
}
fn external_energy(
    seq: &[u8],
    _pair_table: &[i32],
    top_pairs: &[(usize, usize)],
    n: usize,
) -> i32 {
    let mut energy = 0i32;
    for &(i, j) in top_pairs {
        let pi = match pair_index(seq[i], seq[j]) { Some(x) => x, None => continue };
        if is_au_gu(pi) {
            energy += TERMINAL_AU_PENALTY;
        }
        if i > 0 {
            energy += DANGLE5[pi][seq[i - 1] as usize];
        }
        if j + 1 < n {
            energy += DANGLE3[pi][seq[j + 1] as usize];
        }
    }
    energy
}
